chalk-solve/src/clauses.rs

```diff
@@ -127,8 +127,13 @@ pub(crate) fn program_clauses_for_goal<'db, I: Interner>(
     let mut vec = vec![];
     vec.extend(db.custom_clauses());
     program_clauses_that_could_match(db, environment, goal, &mut vec)?;
-    program_clauses_for_env(db, environment, &mut vec);
     vec.retain(|c| c.could_match(interner, goal));
+    vec.extend(
+        db.program_clauses_for_env(environment)
```

Comment:
Why not just move this above and keep the `retain` after the `extend` here? I guess it theoretically uses more memory, but it's cleaner.

Reply:
It would also be a bit slower, as you would have to move elements back after you inserted them.

Reply:
Yeah, I just didn't want to insert hundreds of elements just to remove them immediately afterwards, considering this actually gets called a lot.

Reply:
Theoretically we could also use `.chain().filter()` for everything. But I'm gonna just add a FIXME and land this.